### PR TITLE
Issue #26 - Spell Data

### PR DESCRIPTION
**Pull request by NeptuneRjo**
Added the dnd5e API so that users can choose spells to add to their spell list.

This feature is *mostly* done, but I'm looking for feedback. How does it look? Anything drastic I should change? etc...

### CODE REVIEW

File: src/views/Character5eSheet.js [hunk 1]
```diff
@@ -344,25 +363,77 @@ class Character5eSheet extends SheetView {
...
+            this
+        );
         this.emitter.off('character:save:update', this._updateSaveMods, this);
+        this.emitter.off('character:set', this._addSpellButtonEvents, this);
+        this.emitter.off('character:update:spells', this._updateSpellList, this);
     }
+
+    _addSpellButtonEvents () {
+        const spellButtons = this.shadowRoot.querySelectorAll('[data-level]');
+        Array.prototype.forEach.call(spellButtons, (btn) => {
```
Comment (derikb) [inline src/views/Character5eSheet.js:400]: You can just call `forEach` directly on the result of `querySelectorAll`, no need for `Array.prototype`

File: src/views/Character5eSheet.js [hunk 2]
```diff
@@ -344,25 +363,77 @@ class Character5eSheet extends SheetView {
...
+            this.spellButtons.push(new AddSpell(btn, this.cur_character, this.emitter));
+        });
+    };
+
+    _updateSpellList () {
+        const fields = Array.from(this.shadowRoot.querySelectorAll('[data-name="spells"]'));
```
Comment (derikb) [inline src/views/Character5eSheet.js:406]: Ditto here

File: src/views/Character5eSheet.js [hunk 3]
```diff
@@ -344,25 +363,77 @@ class Character5eSheet extends SheetView {
...
+
+        fields.forEach((el) => {
+            const f = el.getAttribute('data-name');
```
Comment (derikb) [inline src/views/Character5eSheet.js:409]: Data attributes can be accessed as `el.dataset.name`...

Comment (derikb) [reply src/views/Character5eSheet.js:409]: might wanna do that as `const f = el.dataset.name || '';` just in case its not set at all.

File: src/views/Character5eSheet.js [hunk 4]
```diff
@@ -344,25 +363,77 @@ class Character5eSheet extends SheetView {
...
+        
+            if (typeof this.cur_character[f] === 'undefined') {
```
Comment (derikb) [inline src/views/Character5eSheet.js:411]: maybe also check if `f === ''`

File: src/views/Character5eSheet.js [hunk 5]
```diff
@@ -344,25 +363,77 @@ class Character5eSheet extends SheetView {
...
+                return;
+            };
+
+            const subf = el.getAttribute('data-subfield');
```
Comment (derikb) [inline src/views/Character5eSheet.js:415]: ditto here.

File: src/views/Character5eSheet.js [hunk 6]
```diff
@@ -344,25 +363,77 @@ class Character5eSheet extends SheetView {
...
+            const charValue = (subf) ? this.cur_character[f][subf] : this.cur_character[f];
+            const listItems = charValue || [];
+        
+            el.clear();
+       
+            if (listItems.length > 0) {
```
Comment (derikb) [inline src/views/Character5eSheet.js:421]: Probably don't need to check length here if its 0 then forEach just won't do anything.

File: src/views/Character5eSheet.js [hunk 7]
```diff
@@ -344,25 +363,77 @@ class Character5eSheet extends SheetView {
...
+                listItems.forEach((item) => {
+                    if (item.length === 0) {
```
Comment (derikb) [inline src/views/Character5eSheet.js:423]: what is item here, a string? Can probably just check `if (!item)` to cover null, undefined, or empty string.

File: src/views/Character5eSheet.js [hunk 8]
```diff
@@ -493,9 +578,9 @@ class Character5eSheet extends SheetView {
         this._showUnsavedDialog();
     }
     /**
-     * When a save is (un)checked in the UI.
-     * @param {CustomEvent} ev
-     */
+	 * When a save is (un)checked in the UI.
+	 * @param {CustomEvent} ev
+	 */
```
Comment (derikb) [inline src/views/Character5eSheet.js:583]: Not a fan of all these seemingly unnecessary formatting changes. I assume this was some kind of automated thing, maybe you can... revert that?

File: src/components/AddSpell.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+}
+</style>
+`;
+
+class AddSpell {
+    constructor (el, cur_character, emitter) {
+        this.el = el;
+        this.cur_character = cur_character;
+        this.emitter = emitter;
+        this.el.addEventListener('click', this._openSpellModal.bind(this));
```
Comment (derikb) [inline src/components/AddSpell.js:34]: it doesn't cause issues to call an async method in this event listener? Feels like it should have a catch block at least in case the fetch request rejects/throws.

File: src/components/AddSpell.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    };
+    /**
+     * @param {string} spellLevel
+     * @returns array containing the index, name, and url
+     */
+
+    async getSpellData (spellLevel) {
+        const apiUrl = `https://www.dnd5eapi.co/api/spells?level=${spellLevel}&school=illusion&school=abjuration&school=conjuration&school=divination&school=enchantment&school=evocation&school=necromancy&school=psionic&school=transmutation`;
+        const response = await fetch(apiUrl, { method: 'GET' });
+        const { results } = await response.json();
```
Comment (derikb) [inline src/components/AddSpell.js:44]: Probably need error handling here in cause the api is down or something?

File: src/components/AddSpell.js [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        this.emitter = emitter;
+        this.el.addEventListener('click', this._openSpellModal.bind(this));
+    };
+    /**
+     * @param {string} spellLevel
+     * @returns array containing the index, name, and url
+     */
+
+    async getSpellData (spellLevel) {
+        const apiUrl = `https://www.dnd5eapi.co/api/spells?level=${spellLevel}&school=illusion&school=abjuration&school=conjuration&school=divination&school=enchantment&school=evocation&school=necromancy&school=psionic&school=transmutation`;
```
Comment (derikb) [inline src/components/AddSpell.js:42]: what's the deal with this api? free and open? is that list of schools... all the schools? It will work for all classes?

**Review by derikb — COMMENTED** [review]
Haven't loaded it up to see but it looks pretty good.

**Comment by derikb** [comment]
p.s. thanks for working on this.

File: src/components/AddSpell.js (hunk 9 above)
Comment (NeptuneRjo) [reply src/components/AddSpell.js:34]: I couldn't find anything on not being able to use asynchronous functions with event listeners, but I'll add some catches to coincide with the API error handling.

File: src/components/AddSpell.js (hunk 11 above)
Comment (NeptuneRjo) [reply src/components/AddSpell.js:42]: The API is free use. This will get all of the schools for the specific level, and it is getting spells across classes. The only downside to this API is that some spells just aren't available (perhaps from specific books or newer content).

Comment (derikb) [reply src/components/AddSpell.js:42]: Probably the difference between what's SRD and what's proprietary to Hasbro.

**Review by derikb — APPROVED** [review]
Going to merge this, then I'll test it out this weekend.

**Comment by NeptuneRjo** [comment]
Sounds good. Let me know if you need anything from me, I'm going to look through everything again just to make sure I didn't miss anything.
